### ISSUE TITLE
Global Scope

### ISSUE THREAD

**Issue by AndrewCarterUK** [async-interop/event-loop#14]
This is a topic that I've discussed independently with a few people here, and I'm still not sure that I'm 100% convinced by any of the solutions.

I guess the best phrasing of the problem is:

_**"Most PHP application instances will require no more than a single event loop. How can we ensure that this event loop can be easily accessed by code within the application?"**_

From here I believe there are a few options (feel free to add):
1. Straight forward dependency injection. The benefits of this is that it's explicit, you can see immediately whether a class uses the event loop and you have type safety. The negative is that it's repetitive and requires a lot of boiler plate code.
2. Make the event loop global. Have a class that can act as a locator with static properties (or functions) so that code anywhere in the application can retrieve the instance of the event loop.
3. Some hybrid between these, which allows users to choose whether they wish to explicitly require the event loop as a class dependency or just access it statically when they need it.

A big factor at play here is testability. I'm not convinced that anything other than option 1 is a good bet for writing testable code - but I have been told that the other options don't interfere with this. Has anyone got an example of well tested classes that retrieve their instance of the event loop from global scope? I'm happy to be convinced on this point :)

Another question I have is - can anybody think of a use case where a single PHP instance would wish to have multiple event loops operating at any one time?

Interested to hear your thoughts,

Andrew

**Comment by bwoebi** [comment]
Quick start: Nah, please don't go with option 3. That's worse than any of 1 or 2.

---

There _are_ situations where you'd like to have multiple event loops, but there are none where you'd operate two event loops in parallel (that's basically only possible via a busy loop switching between both).
Quick example: https://github.com/amphp/aerys/blob/master/lib/WorkerProcess.php#L22-L29
But this is such a rare case that you can comfortably back up the static instance and put it back after task done.

The use case basically is to have nothing interfering with your code… it's a defacto halting of the global reactor …

So, we still have intent to be able to _swap_ event loops, do something on the new instance and put the old instance back, but not to operate multiple event loops in parallel.

---

For a long time we had the first option in Amp. It worked nice … but it brought literally no benefits except much boilerplate. For the sake of readability we made the overhead (in source) as small as possible, which is just simple functions. `\Amp\once()`. `\Amp\onReadable()`. `\Amp\repeat()`. etc.

Regarding tests, I'm just resetting the global loop to a new instance at the end of each test leaving unclean state.
Also we still could use mocks if needed by just setting the global loop to the mock. - So that's a non-issue.

Another note about _class_. I'm really in favor of using simple global functions (no static class - why a class? does this actually bring any benefit to have it being a _class_?).
For the simple reason that it abstracts the event loop completely away. You don't code to an event loop anymore, but you code to async primitives. [That's by the way also why I proposed #9 - it leaks information about the _loop_ nature of it.] - I schedule something and it just happens. I have to know exactly _nothing_ about the existence of the event loop in the background [well, except that I'm calling run() once … that's the only annoyance left].

KISS. Seriously, just KISS. Choose option 2: global event loop _with functions_.

**Comment by sagebind** [comment]
Here are my thoughts on the matter:
- Defining an interface for event loop I don't think is enough. I want to encourage as many people as possible to write packages in an asynchronous way where applicable, but in most cases, they should be able to do so without depending on any particular loop implementation. For that to work, there needs to be some generic way of retrieving an event loop.
  
  The idea here is that libraries can be written purely on the event loop interface, and it is up to applications to choose an implementation which is then used by all async code in the app.
- Essentially the definition of an event loop is that it is shared. You can't easily run multiple event loops in parallel (on the same thread, anyway :stuck_out_tongue:), so for an event loop to be useful, all parts of an app need to use the same one.
  
  So, there is only one, and it needs to be shared. That's pretty much the definition of a global.
- If we go with (2), I don't have a strong preference between a static class or static functions; only that in either case, `EventLoop::onReadable()` and `EventLoop\onReadable()` read fine in source code, but just `onReadable()` would be really confusing.

So to summarize, I'm in full favor of (2).

**Comment by sagebind** [comment]
@bwoebi The only practical benefit of a static class is that you can have distinct `get()` and `set()` methods for getting and setting the event loop instance, respectively, where you would need a single function to do both operations (unless you used `global`, which wouldn't be awful either).

**Comment by bwoebi** [comment]
Independently of this, I'm not sure where this is heading to - see also #8 ... If we provide functions to directly access a global instance, we also need to provide actual implementations of the interfaces.

@coderstephen hehe, I wanted to use global in Amp too, but we didn't … for esthetic reasons I guess :-D … also not sure if explicit get/set is actually a particular benefit.

Additionally, if we could use a shared namespace for async foundations (eventloop, promises, coroutines), it'll all read nicely as `async\onReadable()`, `async\coroutine()`, `async\Promisor` etc. :-D

**Comment by mbonneau** [comment]
+1 for static (option 2).

With static loop accessors, would the interop group be responsible for implementing that library?

**Comment by sagebind** [comment]
> With static loop accessors, would the interop group be responsible for implementing that library?

Might seem heavy-handed, but I would say yes absolutely. It is the only way to guarantee that packages that want to use interoperable async are using the same accessor, otherwise we're back to square one.

I would provide the loop accessors in the same package as the loop interfaces themselves.

**Comment by sagebind** [comment]
The only thing related is the global-ness, but some of my opinions were inspired by [this particular logging library for Rust](https://doc.rust-lang.org/log/log/index.html). It spread in the Rust community rapidly because it accomplished this idea, where libraries can all use the same logger without knowing the implementation, and the application can choose the logger that is used by every single dependent library.

**Comment by bwoebi** [comment]
I'm not so sure about logging (after all different events might go to different loggers? … that's more a thing for something _package global_ (if such a thing would exist in PHP)) … but that's unrelated^^

Actually, it's not _necessary_ to have these accessors. … the important part is that things arrive in the same and unique global loop.

**Comment by sagebind** [comment]
Yeah, I'm not certain that the design they chose is best for _logging_ necessarily, but said design I do think would work extremely well for event loops.

**Comment by kelunik** [comment]
There's only (2), all other options are non-options.
- (1) Users will start using lazy injection, constructing another reactor otherwise. Having two reactors leads to subtle bugs. We had that in Amp, never again.
- (3) Worst of all options, leads to the same issues as (2).

I don't really care if we choose functions or static methods, but I think I prefer functions, just for the sake of simplicity and the ability to alias namespaces like

``` php
use Async\Interop\EventLoop as Loop;

Loop\onReadable(...);
```

**Comment by AndrewCarterUK** [comment]
I'm a strong believer that this group should not seek to provide implementations of any of the standards that it produces. The intention to do so would get a lot of resistance in the FIG (it would never pass), and there are good reasons for this. I'd prefer to keep that particular discussion in #8, and I can expand on this if people want. The gist is that implementations require a far higher level of maintenance and are much less inviting to couple to than an interface.

Back onto the scope debate, it seems like 2 is the winning option (I thought it might be).

I could only live with functions if they were a proxy for a driver that was configured. As an example:

``` php
use Async\Interop\EventLoop;
use SomeAsyncProject\LibEventLoop;

$loop = new LibEventLoop();
EventLoop\set($loop);

// elsewhere...

EventLoop\onReadable(...);

//

EventLoop\run();
```

However, I think the version of this using a static class as a proxy to an underlying driver (below) is superior.

``` php
use Async\Interop\EventLoop;
use SomeAsyncProject\LibEventLoop;

$loop = new LibEventLoop();
EventLoop::set($loop);

// elsewhere...

EventLoop::onReadable(...);

//

EventLoop::run();
```

I think a static class is superior as it has the benefit of encapsulation, and can store the underlying driver as a static property:

``` php
class EventLoop
{
    private static $driver;

    public static function set(EventLoopInterface $driver)
    {
        self::$driver = $driver;
    }

    public static function get()
    {
        return self::$driver;
    }

    // ...
}
```

Using functions, there's no way to do this without creating an exposed global variable or a single function that you use to both get and set the driver ([as AMP have done](https://github.com/amphp/amp/blob/master/lib/functions.php#L11)). I don't really like either of those options, the first loses type safety and the second breaks SRP.

On the other hand, I can't see any reason not to use a static class (except for having to type an extra character).

Interested to hear thoughts :)

**Comment by AndrewCarterUK** [comment]
Correction, you can use functions without breaking SRP - as you can do this:

``` php

function get()
{
    return loop();
}

function set(EventLoop $loop)
{
    return loop($loop);
}

// some internal api:
function loop(EventLoop $assign = null)
{
    static $loop;

    if (null !== $assign) {
        $loop = $assign;
    }

    return $loop;
}
```

Still, this feels like a messy solution to a problem that static classes solve better.

**Comment by bwoebi** [comment]
@AndrewCarterUK Regarding get() and set(), I had the following discussion: https://github.com/async-interop/event-loop/pull/13#discussion_r54335136
which would render this redundant and it'd be a purely internal construct (no public facing set).

**Comment by AndrewCarterUK** [comment]
@bwoebi - How would you set the default event loop without `set()`?

**Comment by sagebind** [comment]
> I'm a strong believer that this group should not seek to provide implementations of any of the standards that it produces.

@AndrewCarterUK I think I misunderstood; what implementation are you referring to here? Sorry if it was already clarified.

**Comment by AndrewCarterUK** [comment]
@coderstephen I'm referring to any implementation of the `EventLoopInterface` (or whatever it ends up being named). That shouldn't be our concern, this group should only focus on the interface.

@bwoebi - I re-read, are you suggesting that all asynchronous code should be run within a `with()` call?

**Comment by bwoebi** [comment]
@AndrewCarterUK yes I am suggesting a `with()` call for all async code, allowing proper scoping.

**Comment by sagebind** [comment]
I think that, if we were to choose a `get()` and `set()` approach, that Andrew's comment puts into words much more clearly than what I was babbling about a couple days ago on why I prefer the static class approach. Of course, I'm not totally against functions.

**Comment by bwoebi** [comment]
I just feel that `set()` is a bit weird … it sounds like something allowing setting at any time. A set-once function is … well … bit weird to me. And `with()` is going around that problem elegantly IMO.

**Comment by AndrewCarterUK** [comment]
@bwoebi - I'm not opposed to that. I agree that `with()` is more elegant as it's very clear where the scope of the event loop starts and finishes. Would be interested to hear other opinions.

**Comment by sagebind** [comment]
The alternative is what Bob is suggesting: a single `with()` method. To reiterate in this thread, looks like this:

``` php
EventLoop\with(new MyLoop(), function () {
    // do some async code
});
```

Only one function is required here, but all asynchronous code _must_ be run inside the above closure. Suggested advantages are
1. Stack trace for async code will bubble up to the given closure.
2. The loop dependency is specified in the same place that it is run.

**Comment by sagebind** [comment]
@bwoebi Perhaps give the functions better names, like `init()`?

``` php
EventLoop::init(new MyLoop());

// do some async code

EventLoop::run();
```

**Comment by sagebind** [comment]
I feel that this system is akin to the way stream wrappers are set up, class autoloaders are registered, and the like. The main thing I'm not sure that I _do_ like is that creating the event loop has to be in the same place as the beginning of the program; or rather before since it may very well be laid out like this:

``` php
use Async\Interop\EventLoop;
use Vendor\Loop;

// Do setup work first, which may possibly look like this for some vendor.
$factory = new Loop\LoopFactory();
$loop = $factory->createLoop();

// Now do the "real" program in a closure?
EventLoop\with($loop, function () {
    // Do async stuff in here
    // Which is indented from the above code...
});
```

I dunno, maybe I'm weird (actually, I definitely am!), but that indenting situation looks strange to me.

_(sorry for many subsequent posts)_

**Comment by AndrewCarterUK** [comment]
I need to have a play with a few examples, but I can see this approach working.

At the moment, I'm playing with:

``` php
EventLoop::execute(function () {
    // ...

    EventLoop::onReadable($stream, function ($stream) {
        // do handling
    });

    // ...
}, new MyLoop());
```

This would give us a proxy class that looked like:

``` php
class EventLoop
{
    private static $loop = null;

    public static function execute(callable $block, EventLoopInterface $loop)
    {
        self::$loop = $loop;

        $block();

        self::$loop = null;
    }

    private static function inScope()
    {
        if (null === self::$loop) {
            throw new RuntimeException('Not within the scope of an event loop');
        }
    }

    public static function onReadable($stream, callable $callback)
    {
        self::inScope();

        return self::$loop->onReadable($stream, $callback);
    }

    // ...
}
```

https://3v4l.org/1ZF30

**Comment by bwoebi** [comment]
I actually believe that stream wrappers are something horrible and should be banished from PHP ;-D

@coderstephen The core difference between autoloading and this is that you don't have to ever start it (nothing like run()), that it actually is falling back (you can have multiple registered autoloaders) and that it is truly global.
Don't think you can compare these well (apples and oranges ;-)).

**Comment by bwoebi** [comment]
@AndrewCarterUK yeah, basically this (just that you'd place `self::$loop = null;` inside a `finally` and call `self::$loop->run()` inside it … but for the general idea, yeah) :-)

**Comment by AndrewCarterUK** [comment]
As an amendment to the example I posted, I believe the `run()` call would be implicit and should  also be within the try/finally block of the `execute()` method.

**Comment by AndrewCarterUK** [comment]
The code I posted also doesn't account for embedding event loops within event loops. This is the version I'm playing with now: https://3v4l.org/oSZ80

``` php
class EventLoop
{
    private static $loop = null;

    public static function execute(callable $code, EventLoopInterface $loop)
    {
        $previousLoop = self::$loop;

        self::$loop = $loop;

        try {
            $code();

            self::run();
        } finally {
            self::$loop = $previousLoop;
        }
    }

    private static function inScope()
    {
        if (null === self::$loop) {
            throw new RuntimeException('Not within the scope of an event loop');
        }
    }

    public static function run()
    {
        self::inScope();

        self::$loop->run();
    }

    public static function onReadable($stream, callable $callback)
    {
        self::inScope();

        return self::$loop->onReadable($stream, $callback);
    }

    // ...
}
```

Allowing for:

``` php
EventLoop::execute(function () {
    // ...

    EventLoop::onReadable($stream, function ($stream) {
        // do handling
    });

    // ...
}, new MyLoop());
```

**Comment by AndrewCarterUK** [comment]
@trowski @bwoebi @assertchris - Can I get your opinions on the discussion that has happened here?

The proposals are quite structural and I think we've reached a bit of a fork in the road until we can answer this question.

**Comment by sagebind** [comment]
Reflecting the discussion in #5, I'd like to make a point about extensibility. As @AndrewCarterUK pointed out, offering additional functions other than the ones defined in the standard poses a problem for interoperable code. If a specific library provides the capability to listen for Windows process messages, for example, they might add an `onWindowMessage()` method. Now proprietary methods aren't bad, but we can't guarantee they exist in the standard.

The solution would be to allow arbitrary access to the event loop instance. For example:

``` php
EventLoop::execute(function () {
    EventLoop::get()->onWindowMessage(function ($message) {
        // do handling
    });
}, new MyWindowsLoop());
```

We can't guarantee that `onWindowMessage()` exists, but the application or library presumably knows already what specific loop they are compatible with, and so they use that interface. The solution would _not_ be to try and extend `EventLoop` to have `onWindowMessage()`. It wouldn't be necessary.

**Comment by AndrewCarterUK** [comment]
@coderstephen - It's my honest opinion that we shouldn't try to encourage extensibility in any form. If you need functionality that isn't offered by the event loop - you **need** a different event loop. There's no point making any attempt to couple to a generic event loop - you **need** to couple to a specific event loop and this standard isn't useful to you.

There's nothing wrong with trying to solve the 95% use case, and that only.

**Comment by sagebind** [comment]
I'm only suggesting that the interface can't promise anything other than the standard, but we should probably allow access to the actual event loop instance for the reasons I described.

**Comment by AndrewCarterUK** [comment]
I guess that would allow people to hack the event loop, but I'm not convinced it's a good idea to encourage. We should make sure if we refer to this behaviour in examples we do it like this:

``` php
function some_code()
{
    $driver = EventLoop::get();

    if ($driver instanceof MyWindowsLoop) {
        $driver->onWindowMessage(function ($message) {
            // do handling
        });
    }
}

EventLoop::execute(function () {
    // do stuff

   some_code();

   // do some other stuff
}, new MyWindowsLoop());
```

That is, if there is any chance that a different event loop should be used, they should be coupling directly to that implementation rather than the standard. The only use case I can see for this, is if there is an optional feature that is used when present.

**Comment by sagebind** [comment]
To sum up your reply, yes to all of it. :wink:

**Comment by sagebind** [comment]
Here's what I propose: either a static function structure, or a static class structure. The static function structure:

``` php
function get()
{
    global $__eventLoop;
    return $__eventLoop;
}

function run(callable $closure, EventLoop $loop)
{
    global $__eventLoop;
    $prev = $__eventLoop;
    $__eventLoop = $loop;
    try {
        $closure();
    } finally {
        $__eventLoop = $prev;
    }
}

// Other wrapper functions around EventLoop methods.
```

I really wish we had C-style statics that were file-level visible, but oh well. Or the static class equivalent:

``` php
class EventLoop
{
    private static $loop;
    public static function get()
    {
        return self::$loop;
    }
    public static function run(callable $closure, EventLoopDriver $loop)
    {
        $prev = self::$loop;
        self::$loop = $loop;
        try {
            $closure();
        } finally {
            self::$loop = $prev;
        }
    }
}
```

**Comment by bwoebi** [comment]
@AndrewCarterUK The problem with coupling directly is that we'd need to persist it separately again and get things messed up when stacking etc.

Hence the event loop should be accessible in any way from the exterior.

**Comment by kelunik** [comment]
We might need also some mechanism to store global state bound to a reactor. This can be used for things like a default DNS resolvers or filesystem drivers.

Since we have I/O bound applications, most operations will have something to do with networking operations. Most of these need DNS, so it might make sense to have a global accessor for DNS just like for the event loop itself.

Otherwise it's possible to pass it optional and create a new one if no resolver is specified. Multiple instance aren't a problem for DNS, just for the event loop.

**Comment by bwoebi** [comment]
I think we've determined that we're going to use global scope ... This issue can be closed, I think.

@kelunik May you please open a separate issue for that specific point of global state storing?

**Comment by AndrewCarterUK** [comment]
As per usual, I'll close and if anyone has anything to add we can reopen
